### PR TITLE
Support for uncontracted and contracted braille input.

### LINKED ISSUE #2439: Support for entry of contracted braille via a braille keyboard
**Reported by jteh on 2012-06-12 06:11**
(Spun off #808.)

Supporting input of contracted braille is much more complicated than computer braille, as we need to determine when to translate a chunk of input, the user experience before translation of a chunk occurs, etc.
**Blocked by #808**

### LINKED ISSUE #6054: add backspace key binding to refreshabraille18
I have noticed that the refreshabraille18 does not have any binding for the backspace key, however it does have enter and the arrows. I'm wondering if you could make it possible for say b7+b9, or dot7 chord to be used as a backspace such as on IOS.

### LINKED ISSUE #6113: Braille settings: Sort output/input braille tables list
Hi,

Currently, output/input braille tables list shows the list of braille tables based on the order found in braille.TABLES tuple. Although it allows one to reassign braille tables via a simple index operations, it might be more desirable to allow users to go through all tables in sorted order (for example, going from English to French as opposed to English and then Spanish).

To achieve this, a new dictionary named "tables" will be used to replace two lists, with the key being the name of the table and value being filenames. For example, BrailleSettingsDialog.tables["English (U.S.) grade 1"] points to "en-us-g1.ctb".

Surrounding implementations:
1. When loading table names, sort tables.keys.
2. For OK button handler, look up table names.GetStringSelection and assign the value found for this key.
3. If this is working for output tables, input tables presentation will be modified accordingly.

Possible issues/questions:
- How about non-English locales? We can take advantage of the fact that Python is aware of Unicode, hence one can sort using this.
- Is current approach acceptable? When one wishes to look at tables loaded with no regard to ordering, yes, as current method displays table names in the order they are declared in the braille module. The sorting part is handy if one wishes to browse through tables in a sorted order, similar to list of braille displays and synthesizers.
- GetStringSelection versus GetSelection: GetStringSelection is better in that we can use it as the key to the dictionary. Using GetSelection means looking up the string once more based on its index.

Thanks.

### LINKED ISSUE #6935: Upgrade to Liblouis 3.2.0
Hi,

Liblouis 3.1.0 has been released. This release includes tons of changes to back-translation submitted by @JCSTeh (a great help for #2439), braille table improvements, and couple new tables such as Chinese (simplified) table and enhanced Greek and Turkish tables.

Thanks.

### PR DESCRIPTION
**Pull request by jcsteh**
Beyond the main code in brailleInput to support uncontracted/contracted input, this includes the following changes:
- The list of braille tables has been moved out of the braille module into a separate brailleTables module. Braille tables are now added with a function rather than directly adding them to the data structure. Aside from being necessary in order to specify and check whether a table is contracted, this also makes the data about tables more extensible in future.
- As the data structure for braille tables has now changed and is no longer ordered, this was a good opportunity to sort the list of tables alphabetically when displaying them to the user.
- brailleInput is now notified when reverting config or changing config profiles. This is necessary because brailleInput now maintains some state when the input table is changed.
- brailleInput is now initialised before braille at startup. This is because braille depends on brailleInput to get the currently untranslated input.
- Dot7 and dot8 are now universally bound to braille input specific scripts for erase and enter. Any braille display drivers that had bindings for backspace/enter for braille input have been adjusted accordingly.
- In the User Guide, a "Braille" section has been added above "Application Specific Features". The "Braille Control Types and States" section has been moved to a sub-section of this, and a sub-section on Braille Input has been added.

Fixes #2439. Fixes #6054. Fixes #6113. Fixes #6935.

What's New entries:

In New Features:

```
- You can now type in both contracted and uncontracted braille on a braille display with a braille keyboard. See the Braille Input section of the User Guide for details. (#2439, #6054)
```

In Changes:

```
- The output and input table lists in the Braille Settings dialog are now sorted alphabetically. (#6113)
- Updated liblouis braille translator to 3.2.0. (#6935)
```

### CODE REVIEW

**Comment by jcsteh** [comment]
@feerrenrut, would you mind reviewing this please? Note that I've tried to explain some of the more obscure changes in the description of this pull request.

File: source/brailleInput.py
Comment (feerrenrut) [inline source/brailleInput.py]: 0x2800? Can this be changed to a named constant?

File: source/globalCommands.py
Comment (feerrenrut) [inline source/globalCommands.py]: missing translators comment

File: source/gui/settingsDialogs.py
Comment (feerrenrut) [inline source/gui/settingsDialogs.py]: `wx.ID_ANY` is unnecessary, I think this might conflict with `next`

File: source/speech.py
Comment (feerrenrut) [inline source/speech.py]: Can this be called twice before `number` has had a chance to decrement to 0?

File: source/speech.py
Comment (feerrenrut) [inline source/speech.py]: I think this logic could be cleaned up so the condition does not need to be repeated. It would have the implication that `_suppressSpeakTypedCharactersData` exists for longer, and number is allowed to reach zero.

eg:
```
if _suppressSpeakTypedCharactersData:
	...
	timeOk = time.time() - supTime <= 0.1

	suppress = number > 0 and timeOk
	if suppress:
		_suppressSpeakTypedCharactersData = (--number, supTime)
	else:
		_suppressSpeakTypedCharactersData = None
else:
	suppress = False
if not suppress and config.conf["keyboard"]["speakTypedCharacters"] and ord(ch)>=32:
	speakSpelling(realChar)
```

File: source/speech.py
Comment (feerrenrut) [inline source/speech.py]: I'm guessing 32 is a limit for non-printable characters? A named constant would make this more readable.

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: source/brailleDisplayDrivers/brailleNote.py
Comment (jcsteh) [inline source/brailleDisplayDrivers/brailleNote.py]: Weird. This doesn't even show up in diffs from command line git. I reckon this is a GitHub bogus thingy. :(

File: source/brailleInput.py
Comment (jcsteh) [inline source/brailleInput.py]: Yes. In English contracted braille, "alw" is the contraction for "always". However, we don't know it will produce "always" until after the word is complete. So, as the user types, they will hear "a l w". I've tried to clarify this comment a bit, but it's a bit tricky to explain.

File: source/speech.py
Comment (jcsteh) [inline source/speech.py]: Quite true. However, as that's very much unrelated code that existed previously (I just added an extra condition), I'd prefer to address this somewhere else.

File: source/brailleInput.py
Comment (jcsteh) [inline source/brailleInput.py]: No reason at all. I'm a moron. :)

**Comment by jcsteh** [comment]
@feerrenrut, I'm done with this batch of changes. I addressed your code review comments plus some other stuff noted below. Would you mind taking a look again?

@derekriemer reported a bug to me via IRC. STR:

1. Enable input with UEB grade 2.
2. Type "b" and press space.
3. Press dot 6 twice.
4. Press backspace.

This causes an exception. I've fixed this in 0b643d69.

I've also addressed a couple of other issues @derekriemer reported in recent comments on #2439.

**Comment by jcsteh** [comment]
> @feerrenrut, I'm done with this batch of changes.

Actually, I just pushed one more commit to mask characters in protected fields (e.g. password fields). Thanks @derekriemer for catching!

**Review by feerrenrut — APPROVED** [review]

**Comment by jcsteh** [comment]
Holding this back from the 2016.4 release, as there are some major outstanding issues:

* [ ] Investigate/fix effect of undefined dots on subsequent translation. For example, with  UEB g2, ⠰⠁⠃ gets translated to "about" instead of "ab" with noUndefinedDots. Without that mode, you get "\56/ab".
* [ ] When there is untranslated input, routing braille outside of the untranslated input should clear the untranslated input.
* [ ] Further UEB back-translation fixes from APH.

**Comment by jcsteh** [comment]
Possibly a duplicate of (or at least related to) #6855.
CC @MichaelDCurran.

**Comment by josephsl** [comment]
Hi,

In Liblouis 3.2.0, Greek table name has changed from gr-gr-g1.utb to el.ctb. Thanks.

**Comment by jcsteh** [comment]
@feerrenrut, this is finally ready for another round of review. There have been a lot of changes and quite a long time since you last reviewed it, so you may want to just review from scratch. In case it's helpful, though, the changes since your last review are [5ff07e5b..ff6f627a](https://github.com/nvaccess/nvda/pull/6449/files/5ff07e5b..ff6f627a).

File: source/brailleTables.py
Comment (feerrenrut) [inline source/brailleTables.py]: This docstring should also have `output` and `input`

File: source/speech.py
Comment (feerrenrut) [inline source/speech.py]: Could you replace the value `32` with a name please? I'm not really sure what this is testing for

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: source/brailleTables.py
Comment (derekriemer) [inline source/brailleTables.py]: update copyright year

File: source/brailleTables.py
Comment (derekriemer) [inline source/brailleTables.py]: test test

**Review by feerrenrut — APPROVED** [review]

**Comment by jcsteh** [comment]
@feerrenrut, can you please review the latest commit? Sorry. :(

File: source/brailleInput.py
Comment (feerrenrut) [inline source/brailleInput.py]: Where is this called?

Comment (feerrenrut) [inline source/brailleInput.py]: Perhaps a docstring on this to explain what its for? Also, it would be nice to see a couple of examples of what `identifier` might be.

**Review by feerrenrut — CHANGES_REQUESTED** [review]

**Review by feerrenrut — APPROVED** [review]

**Comment by derekriemer** [comment]
fixes https://github.com/nvaccess/nvda/issues/6956

**Comment by derekriemer** [comment]
I'm amazed at how many tickets this autoclosed. What a huge chunk of work.

**Pull request merged**
